lib/execution.go

```diff
@@ -0,0 +1,562 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lib
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/loadimpact/k6/stats"
+	"github.com/pkg/errors"
+)
+
+// An ExecutionScheduler is in charge of initializing executors and using them
+// to initialize and schedule VUs created by a wrapped Runner. It decouples how
+// a swarm of VUs is controlled from the details of how or even where they're
+// scheduled.
+//
+// The core/local execution scheduler schedules VUs on the local machine, but
+// the same interface may be implemented to control a test running on a cluster
+// or in the cloud.
+//
+// TODO: flesh out the interface after actually having more than one
+// implementation...
+type ExecutionScheduler interface {
+	// Returns the wrapped runner. May return nil if not applicable, eg.
+	// if we're remote controlling a test running on another machine.
+	GetRunner() Runner
+
+	// Return the ExecutionState instance from which different statistics for the
+	// current state of the runner could be retrieved.
+	GetState() *ExecutionState
+
+	// Return the instances of the configured executors
+	GetExecutors() []Executor
+
+	// Init initializes all executors, including all of their needed VUs.
+	Init(ctx context.Context, engineOut chan<- stats.SampleContainer) error
+
+	// Run the ExecutionScheduler, funneling the generated metric samples
+	// through the supplied out channel.
+	Run(ctx context.Context, engineOut chan<- stats.SampleContainer) error
+
+	// Pause a test, or start/resume it. To check if a test is paused, use
+	// GetState().IsPaused().
+	//
+	// Currently, any executor, so any test, can be started in a paused state.
+	// This will cause k6 to initialize all needed VUs, but it won't actually
+	// start the test. Later, the test can be started for real be
```

Comment:
Typo: "for real be"

Sidenote: should I report typos at this stage? I noticed a couple of others, but not sure if it would be relevant or just noise.

Reply:
re:Sidenote: yes, it is unlikely(unless huge problems are found) that this PR will change a lot code wise. We might reorganise things and I hope we will refractor some parts to functions/methods ;) but IMHO we will push it, so if you find typos, now is a good time to report them

Reply:
FYI in order to avoid clutter with "typo" comments here, I pushed these small fixes to a branch based on `new-schedulers`: https://github.com/loadimpact/k6/tree/fix/1007-minor-issues

So I'll resolve the current comments and push there if I find new issues.